# **Task \#2**: Machine Learning MC886/MO444
##**Classical Machine Learning Techniques**##

In [1]:
print('Marcelo Antunes Soares Fantini' + ' RA 108341')
print('Rubens de Castro Pereira' + ' RA 217146') 

Marcelo Antunes Soares Fantini RA 108341
Rubens de Castro Pereira RA 217146


## Objective:

To explore **Linear Regression** and **Logistic Regression** alternatives and come up with the best possible model for the problems, at the same time avoiding overfitting (and also underfitting). In this work, we will train a model that use medical records from patients to predict patient survival.

## Dataset

The purpose of this dataset is to predict the patients' survival from medical records. These data is available in [Google Sheets](https://docs.google.com/spreadsheets/d/18-PlYBnwD7yJoyGJOsaB5Wydc2iTG9sSnbuuV4t4efE/edit?usp=sharing), consisting in 299 patients with heart failure collected in 2015. 

**Data Dictionary**:

- **AGE**: Age of the patient (years);

- **ANAEMIA**: Decrease of red blood cells or hemoglobin (boolean);

- **HIGH BLODD PRESSURE**: If the patient has hypertension (boolean);

- **CREATININE PHOSPHOKINASE (CPK)**: Level of the CPK enzyme in the blood (mcg/L);

- **DIABETES**: If the patient has diabetes (boolean);

- **EJECTION FRACTION**: Percentage of blood leaving the heart at each contraction (percentage);

- **PLATELETS**: Platelets in the blood (kiloplatelets/mL);

- **SEX**: Woman or man (binary);

- **SERUM CREATININE**: Level of serum creatinine in the blood (mg/dL);

- **SERUM SODIUM**: Level of serum sodium in the blood (mEq/L);

- **SMOKING**: If the patient smokes or not (boolean);

- **TIME**: Follow-up period (days);

- **DEATH EVENT**: If the patient deceased during the follow-up period (boolean);

More information about the dataset: *Chicco, D., Jurman, G. Machine learning can predict survival of patients with heart failure from serum creatinine and ejection fraction alone. BMC Med Inform Decis Mak 20, 16 (2020). https://doi.org/10.1186/s12911-020-1023-5*

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt

df = pd.read_csv('data/heart_dataset.csv').drop(columns=['time'])
df.head(5)

age anaemia  creatinine_phosphokinase diabetes  ejection_fraction  \
0  75.0      No                     582.0       No               20.0   
1  55.0      No                    7861.0       No               38.0   
2  65.0      No                     146.0       No               20.0   
3  50.0     Yes                     111.0       No               20.0   
4   NaN     Yes                     160.0      Yes               20.0   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium     sex  \
0                  1.0  265000.00               1.9         130.0  Female   
1                  0.0  263358.03               1.1         136.0  Female   
2                  0.0  162000.00               1.3         129.0  Female   
3                  0.0  210000.00               1.9         137.0  Female   
4                  0.0  327000.00               2.7         116.0    Male   

  smoking  DEATH_EVENT  
0      No            1  
1      No            1  
2     Yes            1  
3      No            1  
4      No            1

### Analysis


(0.5 points) This part of the assignmente aims to analyze the dataset and preprocess the data for the models. To do so, a fews things to be considered: 


*   Are there any outliers?
*   Are there missing values?
*   How will you handle categorical variables?

In [2]:
# TODO: Preprocessing Dataset.

# looking for outliers in dataset
df.describe()




age  creatinine_phosphokinase  ejection_fraction  \
count  295.000000                297.000000         297.000000   
mean    60.946895                584.801347          38.003367   
std     11.896148                972.874154          11.805432   
min     40.000000                 23.000000          14.000000   
25%     51.000000                118.000000          30.000000   
50%     60.000000                250.000000          38.000000   
75%     70.000000                582.000000          45.000000   
max     95.000000               7861.000000          80.000000   

       high_blood_pressure      platelets  serum_creatinine  serum_sodium  \
count           297.000000     296.000000        297.000000    298.000000   
mean              0.353535  263387.475405          1.397542    136.620805   
std               0.478874   98295.456485          1.036956      4.419177   
min               0.000000   25100.000000          0.500000    113.000000   
25%               0.000000  211750.000000          0.900000    134.000000   
50%               0.000000  262000.000000          1.100000    137.000000   
75%               1.000000  304000.000000          1.400000    140.000000   
max               1.000000  850000.000000          9.400000    148.000000   

       DEATH_EVENT  
count    299.00000  
mean       0.32107  
std        0.46767  
min        0.00000  
25%        0.00000  
50%        0.00000  
75%        1.00000  
max        1.00000

### Analysis Answers

Question 1) Are there any outliers?

We used the some tips in https://careerfoundry.com/en/blog/data-analytics/how-to-find-outliers/ to help the analysis of dataset 

Evaluating the summary statistics of the dataset for each column:

1) age: no outliers, because the min and max values are around the mean value
2) creatinine_phosphokinase: we have outliers because the max value
   mean: 584.801347	
   min: 23.000000       (3.93% of the mean)
   max: 7861.000000     (1344% of the mean)
3) ejection_fraction: no outliers
   mean: 38.003367
   min: 14.000000       (36.83% of the mean)
   max: 80.000000       (2.10% of the mean)
4) high_blood_pressure: ???????????????
   mean: 0.353535
   min: 0.000000        (0% - this is an measure error??? The person is dieing!!)
   max: 1.000000        (282.85% of the mean)
5) platelets: we have outliers because the max value
   mean: 263387.475405
   min: 25100.000000    (9.53% of the mean)
   max: 850000.000000   (322.71% of the mean)
6) serum_creatinine: no outliers
   mean: 1.397542
   min: 0.500000        (35.77% of the mean)
   max: 9.400000        (6.73% of the mean)
7) serum_creatinine: no outliers
   mean: 1.397542
   min: 0.500000        (35.77% of the mean)
   max: 9.400000        (6.73% of the mean)
8) DEATH_EVENT: no outliers because it's a boolean data 


## Linear Regression

(0.75 points) This part of the assigment aims to predict the level of serum creatinine in the blod based on their medical records. 

*   Do we need to split the data into train, valid and test? How?
*   Do we need to normalize the data? How? The normalization is equal to train, valid and test split?
* **Target value: serum_creatinine**.

Obs: drop the DEATH_EVENT column.

In [ ]:
# TODO: Split data and normalize. You can use scikit-learn libraries.

### Activities

1. (0.5 points) Verify if there is any feature that has low correlation with the target variables. You can use the function [mutual_info_regression](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_regression.html) to check the importance of each feature. 

> *   Do we need all the features to predict the target value?
> *   What happens if we drop the low correlation features?

In [ ]:
from sklearn.feature_selection import mutual_info_regression

# TODO: Feature selection.

> What are the conclusions? (1-2 paragraphs)

2. (2.0 points) Perform Linear Regression. You should implement your solution and compare it with ```sklearn.linear_model.SGDRegressor``` (linear model fitted by minimizing a regularized empirical loss with SGD, http://scikit-learn.org).

In [ ]:
# TODO: MSE Loss. You cannot use scikit-learn!!
def MSE():
  return None

In [ ]:
# TODO: Linear Regression. Implement yout solution. You cannot use scikit-learn, Keras/TensorFlow, or PyTorch libraries!!
class LinearRegression():
  def __init__(self, learning_rate=-1, max_iter=-1):
    self.max_iter         = max_iter
    self.learning_rate    = learning_rate
    self.weights          = None
    self.bias             = None

  def predict(self, X):
    return None

  def fit(self, X, y):
    return None

In [ ]:
# TODO: Linear Regression. You can use scikit-learn libraries.
from sklearn.linear_model import SGDRegressor

> What are the conclusions? (1-2 paragraphs)

3. (0.5 points) Plot the cost function vs. number of epochs in the training/validation set and analyze the model. 

In [ ]:
# TODO: Plot the cost function vs. number of iterations in the training set.

> What are the conclusions? (1-2 paragraphs)

4. (0.75 points) Use different learning rates when optimizing the model. You can use the model that you implemented or the model from scikit-learn.

> If you prefer to use the SGDRegressor from scikit-learn, use the parameter ```learning_rate='constant'``` and change the ```eta0```.

In [ ]:
# TODO: Different learning rates. You can use scikit-learn libraries.

> What are the conclusions? (1-2 paragraphs)

5. (0.5 points) Sometimes, we need some more complex function to make good prediction. Evaluate a [Polynomial Linear Regression](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html) model. 

In [ ]:
# TODO: Complex model. You can use scikit-learn libraries.

> What are the conclusions? (1-2 paragraphs)

6. (0.5 points) Pick **your best model**, based on your validation set, and predict the target values for the test set.

In [ ]:
# TODO: Plot the true label and the predict label

## Logistic Regression

This part of the assigment aims to predict the patients survival based on their medical records. 

*   Do we need to split the data into train, valid and test? How?
*   Do we need to normalize the data? How? The normalization is equal to train, valid and test split?
* **Target value: DEATH EVENT**.

1. (0.5 points) Verify if there is any feature that has low correlation with the target variables. You can use the function [mutual_info_classification](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_classif.html) to check the importance of each feature.

In [ ]:
from sklearn.feature_selection import mutual_info_classif

# TODO: Feature selection

> What are the conclusions? (1-2 paragraphs)

2. (2 points) Perform Multinomial Logistic Regression.

In [ ]:
# TODO: Logistic Regression. You can use scikit-learn libraries.

> What are the conclusions? (1-2 paragraphs)


3. (1.5 points) Another factor that can alter the model result it's if the dataset is balanced:

*   The dataset is balanced ?
*   In this case, which technique is better: *oversampling* or *undersampling*?

Obs: You can use [RandomOverSampling](https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.RandomOverSampler.html) and [RandomUnderSampling](https://imbalanced-learn.org/stable/references/generated/imblearn.under_sampling.RandomUnderSampler.html#imblearn.under_sampling.RandomUnderSampler). But you can try others techniques.

In [ ]:
# TODO: Oversampling. You can use scikit-learn and imblearn libraries.

In [ ]:
# TODO: Undersampling. You can use scikit-learn and imblearn libraries.

> What are the conclusions? (1-2 paragraphs)

4. (0.5 points) Pick **your best model** and plot the confusion matrix in the **test set**. 


In [ ]:
# TODO: Plot the confusion matrix. You can use scikit-learn, seaborn, matplotlib libraries.

> What are the conclusions? (1-2 paragraphs)

## Deadline

Wednesday, April 19, 11:59 pm. 

Penalty policy for late submission: You are not encouraged to submit your assignment after due date. However, in case you do, your grade will be penalized as follows:
- April 20, 11:59 pm : grade * 0.75
- April 21, 11:59 pm : grade * 0.5
- April 22, 11:59 pm : grade * 0.25


## Submission

On Google Classroom, submit your Jupyter Notebook (in Portuguese or English).

**This activity is NOT individual, it must be done in pairs (two-person group).**